### PR TITLE
hash dedup vec

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

Deduplicate and eliminate zero lamport account in accounts hash calculation is slow. 

#### Summary of Changes

Optimize the deduplicate and eliminate zero accounts loop. 

- master

```
sol@dev-equinix-washington-23:~$ grep calculate_accounts_hash_from_storages d1 | tail -n1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "pubkey_bin_search_us|eliminate_zeros_us|min_pk_scan_us|hash_writes_us|skip_duplicates_us|sort_us"
eliminate_zeros_us                  6466915i
pubkey_bin_search_us                14784327i
```

- linked list (https://github.com/solana-labs/solana/pull/33215)
```
sol@dev-equinix-washington-23:~$ grep calculate_accounts_hash_from_storages d1 | tail -n1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "pubkey_bin_search_us|eliminate_zeros_us|min_pk_scan_us|hash_writes_us|skip_duplicates_us"
eliminate_zeros_us                  5390743i
pubkey_bin_search_us                22815254i
```

- this PR 
```
sol@dev-equinix-washington-23:~$ grep calculate_accounts_hash_from_storages d1 | tail -n1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "pubkey_bin_search_us|eliminate_zeros_us|min_pk_scan_us|hash_writes_us|skip_duplicates_us"
eliminate_zeros_us                  5022816i
pubkey_bin_search_us                19616637i
```

We have implemented a linked list version and a vector version (this PR). Above shows the timing comparison from running master/linklist pr/this pr with ledger tool, where both LinkedList and this PR show better performance. We prefer the vector implementation since it is simpler and cleaner. 

Fixes #

### CODE REVIEW

**Comment by HaoranYi** [comment]
@jeffwashington @brooksprumo This PR is ready for review. Thanks!

File: accounts-db/src/accounts_hash.rs [hunk 1]
```diff
@@ -813,30 +813,24 @@ impl<'a> AccountsHasher<'a> {
...
-    ) -> &'b CalculateHashIntermediate {
-        let first_item = first_items[min_index];
-        let key = &first_item;
-        let division_index = first_item_to_pubkey_division[min_index];
+        division_index: usize,
+        offset: usize,
+        key: &Pubkey,
+    ) -> (
+        &'b CalculateHashIntermediate,
+        Option<(&'b Pubkey, usize, usize)>,
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:825]: we may want a type here. 2 usizes - I'm not sure what they are.

File: accounts-db/src/accounts_hash.rs [hunk 2]
```diff
@@ -813,30 +813,24 @@ impl<'a> AccountsHasher<'a> {
...
-    /// or removes the entire pubkey division entries (for `min_index`) if the referenced pubkey is the last entry in the same `bin`
-    ///     removed from: `first_items`, `indexes`, and `first_item_pubkey_division`
     fn get_item<'b>(
-        min_index: usize,
-        bin: usize,
-        first_items: &mut Vec<Pubkey>,
         sorted_data_by_pubkey: &[&'b [CalculateHashIntermediate]],
-        indexes: &mut Vec<usize>,
-        first_item_to_pubkey_division: &mut Vec<usize>,
+        bin: usize,
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:818]: can you reorder the parameters to match what was there before? Or is there a reason to re-order them?

Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:818]: `get_item` was complicated already. a comment would be great.

File: accounts-db/src/accounts_hash.rs [hunk 3]
```diff
@@ -813,30 +813,24 @@ impl<'a> AccountsHasher<'a> {
...
+        &'b CalculateHashIntermediate,
+        Option<(&'b Pubkey, usize, usize)>,
+    ) {
         let division_data = &sorted_data_by_pubkey[division_index];
-        let mut index = indexes[min_index];
+        let mut index = offset;
         index += 1;
-        let mut end;
+        let mut next = None;
+
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:831]: `while index < division_data.len()`?

File: accounts-db/src/accounts_hash.rs [hunk 4]
```diff
@@ -960,25 +946,60 @@ impl<'a> AccountsHasher<'a> {
...
+        // working_set hold the lowest items for each slot_group sorted by pubkey descending (min_key is the last)
+        let mut working_set: Vec<(usize, usize)> = Vec::default();
 
-        // initialize 'first_items', which holds the current lowest item in each slot group
+        // initialize 'working_set', which holds the current lowest item in each slot group
+        // working_set should be initialized in reverse order of slot_groups. Later slot_groups are
+        // processed first. For each slot_group, if the lowest item for current slot group is
+        // already in working_set (i.e. inserted by a later slot group), the next lowest items
+        // in this slot group are searched and checked until either one that is not the
+        // working_set are found, which will then be inserted, or no next lowest items are found.
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:957]: this is enough complexity with few enough side effects that it seems like it should be its own fn.

File: accounts-db/src/accounts_hash.rs [hunk 5]
```diff
@@ -1001,47 +1022,27 @@ impl<'a> AccountsHasher<'a> {
...
-                    std::cmp::Ordering::Greater => {
-                        // this is the new min pubkey
-                        min_pubkey = *key;
-                    }
-                }
-                // this is the new index of the min entry
-                min_index = first_item_index;
-            }
+        let mut new = None;
+        while new.is_some() || !working_set.is_empty() {
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply accounts-db/src/accounts_hash.rs:1028]: ooh, or like you did below:
`while let Some((key, slot_group_index, offset)) = std::mem::take().else ... {`

File: accounts-db/src/accounts_hash.rs [hunk 6]
```diff
@@ -1064,22 +1065,61 @@ impl<'a> AccountsHasher<'a> {
...
+
+                let found = working_set.binary_search_by(|(slot_group_index, offset)| {
+                    let prob = &sorted_data_by_pubkey[*slot_group_index][*offset].pubkey;
+                    key.cmp(prob)
                 });
-                duplicate_pubkey_indexes.clear();
+
+                match found {
+                    Err(index) => {
+                        // found a new new key, insert into the working_set
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1089]: comment that this is the source of theoretical O(n/2)

File: accounts-db/src/accounts_hash.rs [hunk 7]
```diff
@@ -1064,22 +1065,61 @@ impl<'a> AccountsHasher<'a> {
...
-                    Self::get_item(
-                        *i,
-                        pubkey_bin,
-                        &mut first_items,
-                        sorted_data_by_pubkey,
-                        &mut indexes,
-                        &mut first_item_to_pubkey_division,
-                        &binner,
-                    );
+            // looping to add next item to working set
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1068]: I think there's value in putting this in its own fn, too.

File: accounts-db/src/accounts_hash.rs [hunk 8]
```diff
@@ -1064,22 +1065,61 @@ impl<'a> AccountsHasher<'a> {
...
+                        // found a new new key, insert into the working_set
+                        working_set.insert(index, (slot_group_index, offset));
+                        break;
+                    }
+                    Ok(index) => {
+                        let v = &mut working_set[index];
+                        if v.0 > slot_group_index {
+                            // There is already a later slot group that contains this key in the working_set,
+                            // look up again.
+                            let (_item, new_next) = Self::get_item(
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1098]: maybe both of these get_item calls:
`next = Self::get_item(...).1;`
Or, if you make a type out of `get_item`'s return value:
`next = Self::get_item(...).next;`

**Comment by jeffwashington** [comment]
Super exciting to see this making progress! Thank you!

File: accounts-db/src/accounts_hash.rs (hunk 1 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: yeah. done.

File: accounts-db/src/accounts_hash.rs [hunk 9]
```diff
@@ -435,6 +435,13 @@ pub struct AccountsHasher<'a> {
     pub(crate) active_stats: &'a ActiveStats,
 }
 
+/// A struct for the location of an account hash item inside chunked accounts hash slices.
+struct ItemLocation<'a> {
+    key: &'a Pubkey,       // account's pubkey
+    division_index: usize, // chunk index
+    offset: usize,         // offset within a chunk
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: done.

File: accounts-db/src/accounts_hash.rs [hunk 10]
```diff
@@ -435,6 +435,13 @@ pub struct AccountsHasher<'a> {
     pub(crate) active_stats: &'a ActiveStats,
 }
 
+/// A struct for the location of an account hash item inside chunked accounts hash slices.
+struct ItemLocation<'a> {
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs:440]: done.

File: accounts-db/src/accounts_hash.rs [hunk 11]
```diff
@@ -813,59 +824,43 @@ impl<'a> AccountsHasher<'a> {
         (hashes, lamports_total)
     }
 
-    /// returns the item referenced by `min_index`
-    ///   updates `indexes` to skip over the pubkey and its duplicates
-    ///   updates `first_items` to point to the next pubkey
-    /// or removes the entire pubkey division entries (for `min_index`) if the referenced pubkey is the last entry in the same `bin`
-    ///     removed from: `first_items`, `indexes`, and `first_item_pubkey_division`
+    /// Given the item location, return the item in the `CalculatedHashIntermediate` slices and the next item location in the same bin.
+    /// If reaches the end of the `CalculatedHashIntermediate` slice or has exhausted all the accounts in current bin, return None for next item location.
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:828]: grammar: If we reach the end... or If the end is reached...

File: accounts-db/src/accounts_hash.rs [hunk 12]
```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
...
-
-        let len = sorted_data_by_pubkey.len();
-        let mut indexes = Vec::with_capacity(len);
-        let mut first_items = Vec::with_capacity(len);
-        // map from index of an item in first_items[] to index of the corresponding item in sorted_data_by_pubkey[]
-        // this will change as items in sorted_data_by_pubkey[] are exhausted
-        let mut first_item_to_pubkey_division = Vec::with_capacity(len);
-
-        // initialize 'first_items', which holds the current lowest item in each slot group
+    ) -> (Vec<(usize, usize)>, usize) {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:949]: please document what these 3 usizes represent.

File: accounts-db/src/accounts_hash.rs [hunk 13]
```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
...
+        // working_set hold the lowest items for each slot_group sorted by pubkey descending (min_key is the last)
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:950]: 'holds'

File: accounts-db/src/accounts_hash.rs [hunk 14]
```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
...
+        let mut working_set: Vec<(usize, usize)> = Vec::default();
+
+        // initialize 'working_set', which holds the current lowest item in each slot group
+        // working_set should be initialized in reverse order of slot_groups. Later slot_groups are
+        // processed first. For each slot_group, if the lowest item for current slot group is
+        // already in working_set (i.e. inserted by a later slot group), the next lowest items
+        // in this slot group are searched and checked until either one that is not the
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:957]: maybe "not IN the..."?

File: accounts-db/src/accounts_hash.rs [hunk 15]
```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
...
+                        let found = working_set.binary_search_by(|(slot_group_index, offset)| {
+                            let prob = &sorted_data_by_pubkey[*slot_group_index][*offset].pubkey;
+                            k.cmp(prob)
+                        });
+
+                        match found {
+                            Err(index) => {
+                                // found a new new key, so insert into the working_set
+                                working_set.insert(index, (i, offset));
+                                // add_to_working_set = true;
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:981]: `// add_to_working_set = true;`
left over?

File: accounts-db/src/accounts_hash.rs [hunk 16]
```diff
@@ -994,54 +1023,138 @@ impl<'a> AccountsHasher<'a> {
...
+
+            let found = working_set.binary_search_by(|(slot_group_index, offset)| {
+                let prob = &sorted_data_by_pubkey[*slot_group_index][*offset].pubkey;
+                (*key).cmp(prob)
+            });
+
+            match found {
+                Err(index) => {
+                    // found a new new key, insert into the working_set. This is O(n/2) on
+                    // average. Theoretically, this operation could be expansive and maybe
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1065]: expensive? vs expansive?

File: accounts-db/src/accounts_hash.rs [hunk 17]
```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
...
+
                 if let Some(first_pubkey_in_bin) = first_pubkey_in_bin {
-                    let k = hash_data[first_pubkey_in_bin].pubkey;
-                    first_items.push(k);
-                    first_item_to_pubkey_division.push(i);
-                    indexes.push(first_pubkey_in_bin);
+                    let mut key = Some(&hash_data[first_pubkey_in_bin].pubkey);
+                    let mut offset = first_pubkey_in_bin;
+
+                    while let Some(k) = key {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:971]: what do you think of replacing this loop with something like:
```
let next= hash_data[first_pubkey_in_bin...];

            Self::add_next_item(
                &mut next,
                &mut new,
                &mut working_set,
                ...
            );
            if let Some(new) = new {
            working_set.insert(0, new);
            }
            ```

            The logic in here is complicated and mirrors what is in `add_next_item`.

File: accounts-db/src/accounts_hash.rs (hunk 11 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: Done.

File: accounts-db/src/accounts_hash.rs (hunk 12 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: Done.

File: accounts-db/src/accounts_hash.rs (hunk 13 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: done.

File: accounts-db/src/accounts_hash.rs (hunk 14 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: Done.

File: accounts-db/src/accounts_hash.rs (hunk 15 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: Done.

File: accounts-db/src/accounts_hash.rs (hunk 16 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: Done.

File: accounts-db/src/accounts_hash.rs (hunk 17 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: yeah. done.

**Comment by HaoranYi** [comment]
FYI. I added another commit to remove special case "new" from `add_item`.

The new change is that, even the new item is the new min, rather than return it as `new`, it will be added to working_set.
This change will make init working_set code simpler and the loop condition checking simpler.
Since the new min item is inserted in at the end of the vector, the cost of push into and pop from the working will be O(1), It shouldn't affect performance much.

File: accounts-db/src/accounts_hash.rs [hunk 18]
```diff
@@ -960,6 +960,9 @@ impl<'a> AccountsHasher<'a> {
         // already in working_set (i.e. inserted by a later slot group), the next lowest item
         // in this slot group is searched and checked, until either one that is `not` in the
         // working_set is found, which will then be inserted, or no next lowest item is found.
+        // Iterating in reverse order of slot_group will grantee that each slot group will be
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:963]: "grantee" -> guarantee

File: accounts-db/src/accounts_hash.rs [hunk 19]
```diff
@@ -994,54 +1005,138 @@ impl<'a> AccountsHasher<'a> {
...
+                Err(index) => {
+                    // found a new new key, insert into the working_set. This is O(n/2) on
+                    // average. Theoretically, this operation could be expensive and may be further
+                    // optimized in future.
+                    working_set.insert(index, (*slot_group_index, *offset));
+                    break;
+                }
+                Ok(index) => {
+                    let v = &mut working_set[index];
+                    if v.0 > *slot_group_index {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1050]: I think we should have a type in working_set. v.1 > *slot_group_index would also compile but be totally incorrect.
If it was v.slot_group_index > *slot_group_index, the code is more obviously correct by inspection.

File: accounts-db/src/accounts_hash.rs (hunk 18 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs:963]: done.

File: accounts-db/src/accounts_hash.rs (hunk 19 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs:1050]: done.

File: accounts-db/src/accounts_hash.rs [hunk 20]
```diff
@@ -994,54 +1010,152 @@ impl<'a> AccountsHasher<'a> {
...
+                        SlotGroupPointer {
+                            slot_group_index: *slot_group_index,
+                            offset: *offset,
+                        },
+                    );
+                    break;
+                }
+                Ok(index) => {
+                    let pointer = &mut working_set[index];
+                    if pointer.slot_group_index > *slot_group_index {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1073]: looks much better!

File: accounts-db/src/accounts_hash.rs [hunk 21]
```diff
@@ -994,54 +1010,152 @@ impl<'a> AccountsHasher<'a> {
...
+        next: &mut Option<ItemLocation<'b>>,
+        working_set: &mut Vec<SlotGroupPointer>,
+        sorted_data_by_pubkey: &[&'b [CalculateHashIntermediate]],
+        pubkey_bin: usize,
+        binner: &PubkeyBinCalculator24,
+    ) {
+        // looping to add next item to working set
+        while let Some(ItemLocation {
+            key,
+            division_index: slot_group_index,
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:1028]: seems like these are the same thing:
division_index: slot_group_index,

I'm sure my original terminology was confusing.

If they're the same, then one struct should be within the other struct:
```
ItemLocation<'a> {
pubkey,
pointer: WhateverPointer,
```

Thoughts?

File: accounts-db/src/accounts_hash.rs [hunk 22]
```diff
@@ -435,6 +435,24 @@ pub struct AccountsHasher<'a> {
     pub(crate) active_stats: &'a ActiveStats,
 }
 
+/// Pointer to a specific item in chunked accounts hash slices.
+#[derive(Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: done.

**Review by jeffwashington — APPROVED** [review]
lgtm. Thanks for all the changes. I tested this final version on ledger-tool and we're about to test it on pop. I'll also start a mnb validator with this.

**Comment by HaoranYi** [comment]
image

dedup timing comparison from mainent. jw14: master; jw12: this pr ; jw13: linkedlist.

**Comment by HaoranYi** [comment]
> > dedup timing comparison from mainent. jw14: master; jw12: this pr ; jw13: linkedlist.
> 
> What are the big spikes on jw12 and jw13 from?

They are the restart of the validators

**Comment by jeffwashington** [comment]
and note the graph metrics are showing where we see the 2x. On mnb validators at steady state.

**Review by jeffwashington — APPROVED** [review]
lgtm

**Pull request merged**